### PR TITLE
added test cases for lie_group type ode

### PR DESCRIPTION
**Pull request by RituRajSingh878**
#### Brief description of what is fixed or changed

Added test case for the lie_group type ode.
#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
Are you testing coverage here? Which parts of the code are covered by these tests?

This doesn't need a release note.

**Comment by RituRajSingh878** [comment]
> This doesn't need a release note.

Okay

**Comment by RituRajSingh878** [comment]
> Are you testing coverage here? Which parts of the code are covered by these tests?

All the tests can be solved by lie_group method (I find it from the internet). So I am just trying to check that lie_group method is working or not(Raising NotImplementedError).

**Review by certik — APPROVED** [review]

**Comment by oscarbenjamin** [comment]
This PR looks good (mostly - I'll comment on details in a moment)

You can measure the code coverage (if you `pip install coverage pytest pytest-cov`) using
```
$ pytest --cov=sympy.solvers.ode --cov-report=html -m 'not slow' sympy/solvers/tests/test_ode.py
```
(If you also pip install pytest-xdist you can pass e.g. `-n4` to use 4 cores.) That will produce a coverage report in the htmlcov directory. If you open htmlcov/index.html in a web browser and click on the ode module you can see which lines of code were executed while running the tests by the colouring.

What you will see if you do this is that much of the code for the Lie group solvers is untested. That means it probably has many bugs and that many parts might not work - maybe it did work when it was added but untested code can break over time as the rest of the codebase changes. I don't personally understand the Lie group methods well enough to find examples that can test this code properly. That's why it is important to write good tests at the time when code is added - the author of the code probably knows better than anyone else what tests are relevant.

If we want to fix this code then we need to find examples that use the untested parts so we can be sure they work correctly at least in *some* cases. Then we can be much more confident in changing the code.

File: sympy/solvers/tests/test_ode.py [hunk 1]
```diff
@@ -2953,6 +2953,46 @@ def test_lie_group():
...
 
+    eq=diff(f(x),x) + 2*x*f(x) - x*exp(-x**2)
+    sol = Eq(f(x), exp(-x**2)*(C1 + x**2/2))
+    assert sol == dsolve(eq, hint='lie_group')
+    assert checkodesol(eq, sol)[0]
+
+    eq = diff(f(x),x) + f(x)*cos(x) - exp(2*x)
+    sol = Eq(f(x), exp(-sin(x))*(C1 + Integral(exp(2*x)*exp(sin(x)), x)))
+    assert sol == dsolve(eq, hint='lie_group')
+    assert checkodesol(eq, sol)[0]
```
Comment (oscarbenjamin) [inline sympy/solvers/tests/test_ode.py]: Can you make this `assert checkodesol(...) == (True, 0)` and same for the cases below?

**Comment by oscarbenjamin** [comment]
Do all of these examples give the same results without `hint='lie_group'`?

File: sympy/solvers/tests/test_ode.py (hunk 1 above)
Comment (RituRajSingh878) [reply sympy/solvers/tests/test_ode.py]: sure

**Comment by RituRajSingh878** [comment]
> Do all of these examples give the same results without `hint='lie_group'`?

All of them are giving the same result except two test cases in which the result is same but in different formate.

**Comment by RituRajSingh878** [comment]
> Are you testing coverage here? Which parts of the code are covered by these tests?
> 
> This doesn't need a release note.

I am testing the following lines of code -

https://github.com/sympy/sympy/blob/a7988f0e6ebe946e6adb94a10a172c404abbefe5/sympy/solvers/ode.py#L5825

Here lie_group trying to solve partial differential equation and it is hanging some time.Code Above this line is working fine and have no problem in general.

Other line of code that I want to test is - 
https://github.com/sympy/sympy/blob/a7988f0e6ebe946e6adb94a10a172c404abbefe5/sympy/solvers/ode.py#L5835

Here sometime solve is not able to give the solution of complex equation.

Other line is -
https://github.com/sympy/sympy/blob/a7988f0e6ebe946e6adb94a10a172c404abbefe5/sympy/solvers/ode.py#L5846

Sometime separatevars is not able to separate the equation and cause problems.

All parts code of lie_group is working fine except these three places where they may fail and due to this we will not get the solution of ode (If the are solvable by lie group).
This is not the problem of lie_group method. We can improve the lie_method only by improving `pde `,  `solve` and `separatevars` so that they can give the expected result.

For `separatevars ` we can add numeriacl method of separation and it will work most of the time I think.(https://github.com/sympy/sympy/pull/16468)

**Comment by oscarbenjamin** [comment]
The lines you refer to are already covered by the tests. That doesn't mean that the tests cover all outcomes but these are not the big untested bits I was referring to. Measuring coverage I find that this PR increases it by a few lines but there are still some big fully untested bits like the whole of the `lie_heuristic_abaco2_unique_general` and `lie_heuristic_bivariate` functions.

Given that these tests are quite slow to run and they only increase coverage by a few lines I would rather not add all of them unless needed.

Can you work out which ones are actually increasing coverage and also which are solved by some other ODE method? I think if another method solves the same ODE faster then I would rather add that to the tests instead (unless a similar example is already there in which case we don't need it).

Also presumably these tests are all examples that currently work. Did you find others that don't work but should?

**Comment by RituRajSingh878** [comment]
> Also presumably these tests are all examples that currently work. Did you find others that don't work but should?


I find these three equations 

```
In [11]: eq=f(x).diff(x) - (f(x)-x*log(x))**2/x**2 - log(x)                                

In [12]: eq                                                                                
Out[12]: 
                                       2
          d          (-x⋅log(x) + f(x)) 
-log(x) + ──(f(x)) - ───────────────────
          dx                   2        
                              x         

In [13]: eq=x*f(x).diff(x)*log(x)*sin(f(x)) + cos(f(x))*(1 - x*cos(f(x)))                  

In [14]: eq                                                                                
Out[14]: 
                   d                                      
x⋅log(x)⋅sin(f(x))⋅──(f(x)) + (-x⋅cos(f(x)) + 1)⋅cos(f(x))
                   dx                                     

In [15]: eq = x*f(x).diff(x) - f(x)*( x*log(x**2/f(x)) + 2)                                

In [16]: eq                                                                                
Out[16]: 
             ⎛     ⎛  2 ⎞    ⎞     
  d          ⎜     ⎜ x  ⎟    ⎟     
x⋅──(f(x)) - ⎜x⋅log⎜────⎟ + 2⎟⋅f(x)
  dx         ⎝     ⎝f(x)⎠    ⎠     

In [11]: eq=f(x).diff(x) - (f(x)-x*log(x))**2/x**2 - log(x)                                

In [12]: eq                                                                                
Out[12]: 
                                       2
          d          (-x⋅log(x) + f(x)) 
-log(x) + ──(f(x)) - ───────────────────
          dx                   2        
                              x         

In [13]: eq=x*f(x).diff(x)*log(x)*sin(f(x)) + cos(f(x))*(1 - x*cos(f(x)))                  

In [14]: eq                                                                                
Out[14]: 
                   d                                      
x⋅log(x)⋅sin(f(x))⋅──(f(x)) + (-x⋅cos(f(x)) + 1)⋅cos(f(x))
                   dx                                     

In [15]: eq = x*f(x).diff(x) - f(x)*( x*log(x**2/f(x)) + 2)                                

In [16]: eq                                                                                
Out[16]: 
             ⎛     ⎛  2 ⎞    ⎞     
  d          ⎜     ⎜ x  ⎟    ⎟     
x⋅──(f(x)) - ⎜x⋅log⎜────⎟ + 2⎟⋅f(x)
  dx         ⎝     ⎝f(x)⎠    ⎠     


```

All of them are hanging, probably in solving PDE.

**Comment by RituRajSingh878** [comment]
> Can you work out which ones are actually increasing coverage and also which are solved by some other ODE method? I think if another method solves the same ODE faster then I would rather add that to the tests instead (unless a similar example is already there in which case we don't need it).
> 


Sure

**Comment by RituRajSingh878** [comment]
All added test cases are increasing coverage.

**Comment by oscarbenjamin** [comment]
Can you also add the test cases that are failing in a test that is marked as `@XFAIL` and `@SKIP`? (It needs to be skipped if those examples are hanging).

**Comment by oscarbenjamin** [comment]
Looks good. Merging. Thanks!

**Pull request merged**